dynamic_programming/longest_common_subsequence.cpp

```diff
@@ -0,0 +1,208 @@
+	}
+	printf("\n\n");
+	return;
+}
+
+/**
+ * @breif max will find the maximum between 2 values
+ * @param a integer a
+ * @param b integer b
+ * @returns maximum value
+ */
+int max(int a, int b)
+{
+	// if a is greater than b return a
+	if(a > b)
+	{
+		return a;
+	}
+	// else return b
+	return b;
+}
+
+/**
+ * @brief LCS will make 2D-array with optimal solutions of sub-problems
+ * @param string1 String 1
+ * @param m Length of String 1
+ * @param string2 String 2
+ * @param n Length of String 2
+ * @returns 2D-array
+ */
+int** LCS(char *string1,int m, char *string2, int n)
+{
+	// Creating a 2D-array to store the solutions of all sub-problems
+	int **c;
+	
+	// Dynamically allocating memory for 2D-array
+	c = (int **)calloc(m+1, sizeof(int *));
+	for(int i=0; i<=m; i++)
+	{
+		*(c + i) = (int *)calloc(n+1, sizeof(int));
+	}
+	// c[i][j] will have the length of Longest common subsequence when length of string 1 is i and length of string 2 is j
+	
+	// If length of string 2 is 0 then there will be no Longest common subsequence
+	for(int i=1; i<=m; i++)
+	{
+		c[i][0] = 0;
+	}
+	
+	// If length of string 1 is 0 then there will be no Longest common subsequence
+	for(int j=0; j <=n; j++)
+	{
+		c[0][j] = 0;
+	}
+	
+	for(int i=0; i<m; i++)
+	{
+		for(int j=0; j<n; j++)
+		{
+			// If string1[i] == string2[j]
+			if(string1[i] == string2[j])
+			{
+				c[i+1][j+1] = c[i][j]+1;
+			}
+			// If string1[i] != string2[j]
+			else
+			{
+				c[i+1][j+1] = max(c[i][j+1], c[i+1][j]);
+			}
+		}
+	}
+	return c;
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main()
+{
+	/// string1 and string2 are used to store strings in which we have to find LCS
+	char *string1, *string2;                      ///< 2 strings of max size MAX 
+	string1 = (char *)calloc(MAX, sizeof(char));  ///< memory allocation for string 1
+	string2 = (char *)calloc(MAX, sizeof(char));  ///< memory allocation for string 2
+	
+	/// Assigning Values of 'GAC' for string1
+	string1[0] = 'G';
+	string1[1] = 'A';
+	string1[2] = 'C';
+
+	/// Assigning Values of 'AGCAT' for string2
+	string2[0] = 'A';
+	string2[1] = 'G';
+	string2[2] = 'C';
+	string2[3] = 'A';
+	string2[4] = 'T';
+	
+	// lengths of two Strings;
+	int lenOfStr1 = strlen(string1);       /// length of string1
+	int lenOfStr2 = strlen(string2);       /// length of string2
+	
+	int** c = LCS(string1, lenOfStr1, string2, lenOfStr2);   /// To make the 2D-Array
+	int lengthOfLCS = c[strlen(string1)][strlen(string2)];   /// Length of LCS
+	printf("Length of Longest Common Subsequence is %d...\n\n", lengthOfLCS);
+		
+	// Character Array to store Longest common subsequence
+	char a[lengthOfLCS];
+	int index=0;
+	
+	// If there is no Longest common subsequence
+	if (!lengthOfLCS)
+	{
+		printf("There is no Possible LCS...\n\n");
+		return 0;
+	}
+	// If there is a Longest common subsequence then printing them
+	printf("All Possible LCS are : \n\n");
+	allPossibleLCS(c, a, index, lenOfStr1, lenOfStr2, string1, string2);
```

Comment:
please write a separate test function with some tests (including corner cases) to verify the implementation.

Reply:
Also don't forget to release the dynamically allocated memory resources.

Reply:
Can I use some global variables to write the test function?

Reply:
It's not a good practice to use global variables.

Reply:
Thanks for the suggestion @ayaankhan98 !
I'll try my best.